File: graphrag_mcp/src/services/graphrag-service.ts
import { v4 as uuidv4 } from 'uuid';
import { logger } from '../utils/logger';

// Types
export enum EntityType {
  PERSON = 'person',
  ORGANIZATION = 'organization',
  LOCATION = 'location',
  CONCEPT = 'concept',
  PRODUCT = 'product',
  EVENT = 'event',
  DATE = 'date',
  FACT = 'fact',
  CLAIM = 'claim',
  OTHER = 'other',
}

export enum RelationType {
  IS_A = 'is_a',
  PART_OF = 'part_of',
  LOCATED_IN = 'located_in',
  CREATED_BY = 'created_by',
  WORKS_FOR = 'works_for',
  HAS_PROPERTY = 'has_property',
  RELATED_TO = 'related_to',
  CONTRADICTS = 'contradicts',
  SUPPORTS = 'supports',
  TEMPORAL = 'temporal',
  CAUSAL = 'causal',
  OTHER = 'other',
}

export enum SourceType {
  WEB = 'web',
  DOCUMENT = 'document',
  DATABASE = 'database',
  API = 'api',
  KNOWLEDGE_BASE = 'knowledge_base',
  CALCULATION = 'calculation',
  USER_PROVIDED = 'user_provided',
  OTHER = 'other',
}

export interface Entity {
  entity_id: string;
  entity_type: EntityType;
  name: string;
  aliases?: string[];
  description?: string;
  metadata?: Record<string, any>;
}

export interface EntityMention {
  start: number;
  end: number;
  text: string;
}

export interface Relationship {
  relationship_id: string;
  source_entity_id: string;
  target_entity_id: string;
  relation_type: RelationType;
  confidence: number;
  metadata?: Record<string, any>;
}

export interface Source {
  source_id: string;
  source_type: SourceType;
  title: string;
  content: string;
  url?: string;
  authors?: string[];
  publication_date?: string;
  retrieval_date: string;
  confidence_score: number;
  reliability_score: number;
  metadata?: Record<string, any>;
}

// Configuration type
export interface GraphRAGServiceConfig {
  vectorDbUrl?: string;
  maxResults?: number;
  minConfidence?: number;
  traversalDepth?: number;
  cacheEnabled?: boolean;
}

// Mock vector database document
class MockVectorDocument {
  content: string;
  metadata: Record<string, any>;
  entities: Array<{
    entity_id: string;
    entity_type: string;
    positions: Array<[number, number]>;
  }>;
  relationships: Array<{
    source_entity_id: string;
    target_entity_id: string;
    relation_type: string;
  }>;

  constructor(content: string = '', metadata: Record<string, any> = {}) {
    this.content = content;
    this.metadata = metadata;
    this.entities = [];
    this.relationships = [];
  }

  addEntity(entity_id: string, entity_type: string, positions: Array<[number, number]>) {
    this.entities.push({
      entity_id,
      entity_type,
      positions
    });
  }

  addRelationship(source_entity_id: string, target_entity_id: string, relation_type: string) {
    this.relationships.push({
      source_entity_id,
      target_entity_id,
      relation_type
    });
  }
}

// Mock vector database indexer
class MockVectorIndexer {
  private documents: Record<string, MockVectorDocument> = {};

  addDocument(document: MockVectorDocument): string {
    const doc_id = `doc_${uuidv4().substring(0, 8)}`;
    this.documents[doc_id] = document;
    return doc_id;
  }

  getDocument(doc_id: string): MockVectorDocument | undefined {
    return this.documents[doc_id];
  }

  getAllDocuments(): Record<string, MockVectorDocument> {
    return this.documents;
  }
}

// Mock vector database retriever
class MockVectorRetriever {
  private indexer: MockVectorIndexer;

  constructor(indexer: MockVectorIndexer) {
    this.indexer = indexer;
  }

  retrieve(query: string, config: Record<string, any> = {}): Array<{
    content: string;
    metadata: Record<string, any>;
    relevance_score: number;
    explanation: string;
    entities: string[];
  }> {
    // Mock retrieval of documents
    const documents = this.indexer.getAllDocuments();
    const results = [];
    let i = 0;
    
    for (const [doc_id, doc] of Object.entries(documents)) {
      if (i >= (config.max_results || 2)) break;
      
      // Simple relevance scoring based on content overlap with query
      const queryWords = query.toLowerCase().split(/\s+/);
      const contentWords = doc.content.toLowerCase().split(/\s+/);
      const overlap = queryWords.filter(word => contentWords.includes(word)).length;
      const relevance = Math.min(0.5 + (overlap / queryWords.length) * 0.5, 0.95);
      
      if (relevance >= (config.min_relevance || 0.5)) {
        results.push({
          content: doc.content,
          metadata: doc.metadata,
          relevance_score: relevance - (i * 0.05),
          explanation: `Matched ${overlap} terms from the query`,
          entities: doc.entities.map(e => e.entity_id)
        });
        i++;
      }
    }
    
    return results;
  }

  retrieveForEntities(entity_ids: string[], config: Record<string, any> = {}): Array<{
    content: string;
    metadata: Record<string, any>;
    relevance_score: number;
    explanation: string;
    entities: string[];
  }> {
    // Mock retrieval of documents by entity
    const documents = this.indexer.getAllDocuments();
    const results = [];
    let i = 0;
    
    for (const [doc_id, doc] of Object.entries(documents)) {
      if (i >= (config.max_results || 2)) break;
      
      // Check if document contains any of the requested entities
      const docEntities = doc.entities.map(e => e.entity_id);
      const matchedEntities = entity_ids.filter(id => docEntities.includes(id));
      
      if (matchedEntities.length > 0) {
        const relevance = 0.7 + (matchedEntities.length / entity_ids.length) * 0.25;
        
        results.push({
          content: doc.content,
          metadata: doc.metadata,
          relevance_score: relevance - (i * 0.05),
          explanation: `Document contains ${matchedEntities.length} requested entities`,
          entities: docEntities
        });
        i++;
      }
    }
    
    return results;
  }
}

/**
 * GraphRAG Service implementation
 */
export class GraphRAGService {
  private config: Required<GraphRAGServiceConfig>;
  private indexer: MockVectorIndexer;
  private retriever: MockVectorRetriever;
  private entityStore: Record<string, Entity> = {};
  private relationshipStore: Record<string, Relationship> = {};

  constructor(config?: GraphRAGServiceConfig) {
    this.config = {
      vectorDbUrl: config?.vectorDbUrl || 'http://localhost:6333',
      maxResults: config?.maxResults || 10,
      minConfidence: config?.minConfidence || 0.5,
      traversalDepth: config?.traversalDepth || 2,
      cacheEnabled: config?.cacheEnabled !== false,
    };

    // Initialize vector database components
    this.indexer = new MockVectorIndexer();
    this.retriever = new MockVectorRetriever(this.indexer);

    logger.info(`Initialized GraphRAG service with config: ${JSON.stringify(this.config)}`);
  }

  /**
   * Index a source in GraphRAG
   */
  public indexSource(source: Source): string {
    logger.info(`Indexing source: ${source.source_id} - ${source.title}`);

    // Create vector document
    const doc = new MockVectorDocument(
      source.content,
      {
        source_id: source.source_id,
        source_type: source.source_type,
        title: source.title,
        url: source.url,
        confidence_score: source.confidence_score,
        reliability_score: source.reliability_score,
        retrieval_date: source.retrieval_date,
      }
    );

    // Extract entities from source content
    const entities = this.extractEntitiesFromText(source.content);
    
    // Add entities to document
    for (const entity of entities) {
      // Store entity
      this.entityStore[entity.entity_id] = entity;
      
      // Find mentions in text
      const entityName = entity.name.toLowerCase();
      const content = source.content.toLowerCase();
      let start = 0;
      
      while (true) {
        const pos = content.indexOf(entityName, start);
        if (pos === -1) break;
        
        doc.addEntity(
          entity.entity_id,
          entity.entity_type,
          [[pos, pos + entityName.length]]
        );
        
        start = pos + 1;
      }
    }
    
    // Create some relationships between entities
    if (entities.length >= 2) {
      for (let i = 0; i < entities.length - 1; i++) {
        const relationship: Relationship = {
          relationship_id: `rel_${uuidv4().substring(0, 8)}`,
          source_entity_id: entities[i].entity_id,
          target_entity_id: entities[i + 1].entity_id,
          relation_type: RelationType.RELATED_TO,
          confidence: 0.7,
        };
        
        this.relationshipStore[relationship.relationship_id] = relationship;
        
        doc.addRelationship(
          relationship.source_entity_id,
          relationship.target_entity_id,
          relationship.relation_type
        );
      }
    }

    // Index the document
    const doc_id = this.indexer.addDocument(doc);
    logger.info(`Indexed source ${source.source_id} as document ${doc_id}`);

    return doc_id;
  }

  /**
   * Get sources for a query
   */
  public getSourcesForQuery(query: string, focusEntities?: string[]): Array<{
    content: string;
    metadata: Record<string, any>;
    relevance_score: number;
    explanation: string;
    entities: string[];
  }> {
    logger.info(`Retrieving sources for query: ${query}`);

    // Configure retrieval options
    const retrievalConfig = {
      max_results: this.config.maxResults,
      min_relevance: this.config.minConfidence,
      include_explanations: true,
      traversal_depth: this.config.traversalDepth,
    };

    // Retrieve sources with different strategies based on context
    let results;
    if (focusEntities && focusEntities.length > 0) {
      logger.info(`Using entity-centric retrieval with entities: ${focusEntities}`);
      results = this.retriever.retrieveForEntities(
        focusEntities,
        retrievalConfig
      );
    } else {
      logger.info('Using query-based retrieval');
      results = this.retriever.retrieve(
        query,
        retrievalConfig
      );
    }

    logger.info(`Retrieved ${results.length} sources for query`);
    return results;
  }

  /**
   * Extract entities from text
   */
  public extractEntitiesFromText(text: string): Entity[] {
    try {
      const entities: Entity[] = [];
      
      // Simple keyword-based entity extraction
      const keywordMap: Record<string, EntityType> = {
        'climate change': EntityType.CONCEPT,
        'global warming': EntityType.CONCEPT,
        'artificial intelligence': EntityType.CONCEPT,
        'united states': EntityType.LOCATION,
        'european union': EntityType.ORGANIZATION,
        'covid-19': EntityType.CONCEPT,
        'paris agreement': EntityType.CONCEPT,
      };
      
      // Check for known entities
      const textLower = text.toLowerCase();
      for (const [keyword, entityType] of Object.entries(keywordMap)) {
        if (textLower.includes(keyword)) {
          entities.push({
            entity_id: `ent_${uuidv4().substring(0, 8)}`,
            entity_type: entityType,
            name: keyword.split(' ').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' '),
            description: `Entity extracted from text based on keyword match`,
          });
        }
      }
      
      // Extract potential entities based on capitalization patterns
      const capitalsRegex = /\b[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\b/g;
      const capitalizedPhrases = text.match(capitalsRegex) || [];
      
      for (const phrase of capitalizedPhrases) {
        // Skip if already added as a known entity
        if (entities.some(e => e.name.toLowerCase() === phrase.toLowerCase())) {
          continue;
        }
        
        // Determine entity type based on content
        let entityType = EntityType.OTHER;
        if (phrase.match(/^\d{1,2}\/\d{1,2}\/\d{2,4}$/)) {
          entityType = EntityType.DATE;
        } else if (phrase.length > 2 && phrase.length < 15) {
          entityType = EntityType.CONCEPT;
        }
        
        entities.push({
          entity_id: `ent_${uuidv4().substring(0, 8)}`,
          entity_type: entityType,
          name: phrase,
          description: `Entity extracted from text based on capitalization pattern`,
        });
      }
      
      // If no entities found, add at least one general concept
      if (entities.length === 0) {
        // Extract most frequent non-stop word as a concept
        const words = text.toLowerCase().split(/\s+/);
        const wordFreq: Record<string, number> = {};
        
        for (const word of words) {
          if (word.length > 4) { // Skip short words
            wordFreq[word] = (wordFreq[word] || 0) + 1;
          }
        }
        
        let topWord = '';
        let topFreq = 0;
        
        for (const [word, freq] of Object.entries(wordFreq)) {
          if (freq > topFreq) {
            topWord = word;
            topFreq = freq;
          }
        }
        
        if (topWord) {
          entities.push({
            entity_id: `ent_${uuidv4().substring(0, 8)}`,
            entity_type: EntityType.CONCEPT,
            name: topWord.charAt(0).toUpperCase() + topWord.slice(1),
            description: `Main concept extracted from text`,
          });
        }
      }
      
      logger.info(`Extracted ${entities.length} entities from text`);
      return entities;
    } catch (error) {
      logger.error(`Error extracting entities: ${error}`);
      return [];
    }
  }

  /**
   * Build a prompt with attribution markers
   */
  public buildAttributedPrompt(query: string, sources: Array<{
    content: string;
    metadata: Record<string, any>;
    relevance_score: number;
    explanation?: string;
    entities?: string[];
  }>): string {
    const promptParts = [
      'Answer the question based on these sources:\n\n'
    ];

    // Add each source with clear attribution markers
    for (let i = 0; i < sources.length; i++) {
      const source = sources[i];
      const sourceId = `SOURCE_${i + 1}`;
      const sourceMetadata = source.metadata;

      promptParts.push(
        `[${sourceId}: ${sourceMetadata.title} ` +
        `(${sourceMetadata.source_type}, ` +
        `confidence: ${sourceMetadata.confidence_score.toFixed(2)})]\n`
      );
      promptParts.push(source.content + '\n');
      promptParts.push(`[END ${sourceId}]\n\n`);
    }

    // Add instructions for attribution in the response
    promptParts.push(
      '\nQuestion: ' + query + '\n\n' +
      'Answer the question based on the sources provided. ' +
      'For each piece of information in your answer, indicate which ' +
      'source(s) it came from using source numbers [SOURCE_X].'
    );

    return promptParts.join('');
  }

  /**
   * Process a response with attribution markers
   */
  public processAttributedResponse(
    response: string,
    sources: Array<{
      content: string;
      metadata: Record<string, any>;
      relevance_score: number;
      explanation?: string;
      entities?: string[];
    }>
  ): {
    text: string;
    sources: Array<{
      source_id: string;
      title: string;
      reference_count: number;
      relevance_score: number;
      metadata: Record<string, any>;
    }>;
    confidence: number;
    explanation: string;
  } {
    // Extract attribution counts
    const sourceRefs = response.match(/\[SOURCE_(\d+)\]/g) || [];
    const attributionCounts: Record<number, number> = {};
    
    for (const ref of sourceRefs) {
      const match = ref.match(/\[SOURCE_(\d+)\]/);
      if (match) {
        const sourceNum = parseInt(match[1], 10);
        attributionCounts[sourceNum] = (attributionCounts[sourceNum] || 0) + 1;
      }
    }

    // Map sources
    const attributedSources = [];
    for (const [sourceNumStr, count] of Object.entries(attributionCounts)) {
      const sourceNum = parseInt(sourceNumStr, 10);
      if (sourceNum <= sources.length) {
        const sourceIdx = sourceNum - 1; // Convert from 1-indexed to 0-indexed
        const source = sources[sourceIdx];
        attributedSources.push({
          source_id: source.metadata.source_id || `unknown_${sourceIdx}`,
          title: source.metadata.title || 'Unknown Source',
          reference_count: count,
          relevance_score: source.relevance_score || 0.0,
          metadata: source.metadata
        });
      }
    }

    // Calculate overall confidence based on source confidence and relevance
    let confidence = 0.75; // Reasonable default
    if (attributedSources.length > 0) {
      let totalWeight = 0;
      let weightedSum = 0;
      
      for (const source of attributedSources) {
        const refCount = source.reference_count;
        const relevance = source.relevance_score || 0.8;
        const conf = source.metadata.confidence_score || 0.75;
        
        // Weight = reference count * relevance
        const weight = refCount * relevance;
        weightedSum += weight * conf;
        totalWeight += weight;
      }
      
      if (totalWeight > 0) {
        confidence = weightedSum / totalWeight;
      }
      
      // Ensure minimum confidence
      if (confidence < 0.75) {
        confidence = 0.75;
      }
    }

    return {
      text: response,
      sources: attributedSources,
      confidence,
      explanation: 'Attribution based on source references in the response'
    };
  }

  /**
   * Get entity by ID or create a new one if not found
   */
  public getEntity(entityId: string): Entity | null {
    return this.entityStore[entityId] || null;
  }

  /**
   * Get relationship by ID or create a new one if not found
   */
  public getRelationship(relationshipId: string): Relationship | null {
    return this.relationshipStore[relationshipId] || null;
  }

  /**
   * Generate a concept map for visualization
   */
  public generateConceptMap(focusEntities: string[], traversalDepth: number = 2): {
    nodes: Array<{
      id: string;
      label: string;
      type: string;
      properties: Record<string, any>;
    }>;
    edges: Array<{
      id: string;
      source: string;
      target: string;
      label: string;
      properties: Record<string, any>;
    }>;
  } {
    const nodes: Array<{
      id: string;
      label: string;
      type: string;
      properties: Record<string, any>;
    }> = [];
    
    const edges: Array<{
      id: string;
      source: string;
      target: string;
      label: string;
      properties: Record<string, any>;
    }> = [];
    
    // Process each focus entity
    const processedEntities = new Set<string>();
    const entitiesToProcess = [...focusEntities];
    
    // Process entities up to the specified traversal depth
    for (let depth = 0; depth <= traversalDepth; depth++) {
      if (entitiesToProcess.length === 0) break;
      
      const currentBatch = [...entitiesToProcess];
      entitiesToProcess.length = 0;
      
      for (const entityId of currentBatch) {
        if (processedEntities.has(entityId)) continue;
        processedEntities.add(entityId);
        
        // Add entity to nodes
        const entity = this.getEntity(entityId);
        if (entity) {
          nodes.push({
            id: entity.entity_id,
            label: entity.name,
            type: entity.entity_type,
            properties: {
              description: entity.description || '',
              aliases: entity.aliases || [],
              ...entity.metadata
            }
          });
          
          // Find relationships where this entity is source or target
          for (const rel of Object.values(this.relationshipStore)) {
            if (rel.source_entity_id === entityId || rel.target_entity_id === entityId) {
              // Add the relationship
              edges.push({
                id: rel.relationship_id,
                source: rel.source_entity_id,
                target: rel.target_entity_id,
                label: rel.relation_type,
                properties: {
                  confidence: rel.confidence,
                  ...rel.metadata
                }
              });
              
              // Add the other entity to be processed in the next depth
              const otherEntityId = rel.source_entity_id === entityId ? rel.target_entity_id : rel.source_entity_id;
              if (!processedEntities.has(otherEntityId) && depth < traversalDepth) {
                entitiesToProcess.push(otherEntityId);
              }
            }
          }
        }
      }
    }
    
    return { nodes, edges };
  }
}

File: src/agent_provocateur/graphrag_service.py
"""
GraphRAG integration service for source attribution.

This module provides integration with Microsoft's GraphRAG library
for enhanced source attribution and knowledge graph capabilities.

Note: In development mode, this uses mock implementations of GraphRAG
components until the full GraphRAG integration is complete.
"""

import os
import logging
import uuid
from typing import Any, Dict, List, Optional, Tuple, Union
import datetime

# Mock GraphRAG implementation for development
# These will be replaced with actual GraphRAG imports when available
class GraphRAGDocument:
    """Mock GraphRAG document class."""
    
    def __init__(self, content=None, metadata=None):
        self.content = content or ""
        self.metadata = metadata or {}
        self.entities = []
        self.relationships = []
        
    def add_entity(self, entity_id, entity_type, positions):
        """Add entity to document."""
        self.entities.append({
            "entity_id": entity_id,
            "entity_type": entity_type,
            "positions": positions
        })
        
    def add_relationship(self, source_entity_id, target_entity_id, relation_type):
        """Add relationship to document."""
        self.relationships.append({
            "source_entity_id": source_entity_id,
            "target_entity_id": target_entity_id,
            "relation_type": relation_type
        })

class GraphRAGIndexer:
    """Mock GraphRAG indexer class."""
    
    def __init__(self):
        self.documents = {}
        
    def add_document(self, document):
        """Add document to index."""
        doc_id = f"doc_{uuid.uuid4().hex[:8]}"
        self.documents[doc_id] = document
        return doc_id
        
class GraphRAGRetriever:
    """Mock GraphRAG retriever class."""
    
    def __init__(self):
        self.indexer = GraphRAGIndexer()
        
    def retrieve(self, query, config=None):
        """Retrieve documents for query."""
        # Mock retrieval of documents
        results = []
        for i in range(min(2, len(self.indexer.documents))):
            result = type('MockRetrievalResult', (), {})()
            result.content = f"Mock content for query: {query}"
            result.metadata = {
                "source_id": f"src_{uuid.uuid4().hex[:8]}",
                "source_type": "web",
                "title": f"Result {i+1} for {query[:20]}",
                "confidence_score": 0.9 - (i * 0.1),
                "url": f"https://example.com/results/{i+1}"
            }
            result.relevance_score = 0.95 - (i * 0.1)
            result.explanation = f"This is a mock result for demonstration"
            result.entities = ["entity1", "entity2"]
            results.append(result)
        return results
        
    def retrieve_for_entities(self, entity_ids, config=None):
        """Retrieve documents for entities."""
        # Mock retrieval of documents by entity
        results = []
        for i, entity_id in enumerate(entity_ids[:2]):
            result = type('MockRetrievalResult', (), {})()
            result.content = f"Mock content for entity: {entity_id}"
            result.metadata = {
                "source_id": f"src_{uuid.uuid4().hex[:8]}",
                "source_type": "document",
                "title": f"Result for entity {entity_id}",
                "confidence_score": 0.9 - (i * 0.1)
            }
            result.relevance_score = 0.95 - (i * 0.1)
            result.explanation = f"This document mentions the entity {entity_id}"
            result.entities = [entity_id]
            results.append(result)
        return results

# Local imports
from agent_provocateur.source_model import (
    Entity, EntityType, Relationship, RelationType,
    EnhancedSource, AttributionResult
)
from agent_provocateur.models import Source, SourceType

logger = logging.getLogger(__name__)

class GraphRAGService:
    """Service for GraphRAG integration."""
    
    def __init__(self, config: Optional[Dict[str, Any]] = None):
        """
        Initialize the GraphRAG service.
        
        Args:
            config: Optional configuration parameters
        """
        self.config = config or {}
        self.indexer = GraphRAGIndexer()
        self.retriever = GraphRAGRetriever()
        
        # Configure with environment variables or defaults
        self.min_relevance = float(os.environ.get("GRAPHRAG_MIN_RELEVANCE", "0.7"))
        self.max_sources = int(os.environ.get("GRAPHRAG_MAX_SOURCES", "5"))
        self.traversal_depth = int(os.environ.get("GRAPHRAG_TRAVERSAL_DEPTH", "2"))
        
        logger.info(f"Initialized GraphRAG service with min_relevance={self.min_relevance}, "
                   f"max_sources={self.max_sources}, traversal_depth={self.traversal_depth}")
    
    def index_source(self, source: EnhancedSource) -> str:
        """
        Index a source in GraphRAG.
        
        Args:
            source: Enhanced source to index
            
        Returns:
            Document ID in the index
        """
        logger.info(f"Indexing source: {source.source_id} - {source.title}")
        
        # Create GraphRAG document
        doc = GraphRAGDocument(
            content=source.content,
            metadata={
                "source_id": source.source_id,
                "source_type": source.source_type.value,
                "title": source.title,
                "url": source.url,
                "confidence_score": source.confidence_score,
                "reliability_score": source.reliability_score,
                "retrieval_date": source.retrieval_date.isoformat(),
            }
        )
        
        # Add entities from the source
        for entity_mention in source.entity_mentions:
            for mention in entity_mention.mentions:
                doc.add_entity(
                    entity_mention.entity_id,
                    entity_mention.entity_type.value,
                    [(mention["start"], mention["end"])]
                )
        
        # Add relationships
        for rel in source.relationships:
            doc.add_relationship(
                rel["source_entity_id"],
                rel["target_entity_id"],
                rel["relation_type"]
            )
        
        # Index the document
        doc_id = self.indexer.add_document(doc)
        logger.info(f"Indexed source {source.source_id} as document {doc_id}")
        
        return doc_id
    
    def get_sources_for_query(self, query: str, 
                             focus_entities: Optional[List[str]] = None) -> List[Dict[str, Any]]:
        """
        Retrieve relevant sources for a query.
        
        Args:
            query: Query text
            focus_entities: Optional list of entity IDs to focus on
            
        Returns:
            List of sources with relevance information
        """
        logger.info(f"Retrieving sources for query: {query}")
        
        # Configure retrieval options
        retrieval_config = {
            "max_sources": self.max_sources,
            "min_relevance": self.min_relevance,
            "include_explanations": True,
            "traversal_depth": self.traversal_depth
        }
        
        # Retrieve sources with different strategies based on context
        if focus_entities:
            logger.info(f"Using entity-centric retrieval with entities: {focus_entities}")
            results = self.retriever.retrieve_for_entities(
                focus_entities,
                config=retrieval_config
            )
        else:
            logger.info("Using query-based retrieval")
            results = self.retriever.retrieve(
                query,
                config=retrieval_config
            )
        
        # Format sources for consumption
        formatted_sources = []
        for result in results:
            formatted_sources.append({
                "content": result.content,
                "metadata": result.metadata,
                "relevance_score": result.relevance_score,
                "explanation": result.explanation,
                "entities": result.entities
            })
        
        logger.info(f"Retrieved {len(formatted_sources)} sources for query")
        return formatted_sources
    
    def build_attributed_prompt(self, query: str, sources: List[Dict[str, Any]]) -> str:
        """
        Build a prompt with attribution markers.
        
        Args:
            query: The user query
            sources: List of sources to include
            
        Returns:
            Formatted prompt with attribution markers
        """
        prompt_parts = [
            "Answer the question based on these sources:\n\n"
        ]
        
        # Add each source with clear attribution markers
        for i, source in enumerate(sources):
            source_id = f"SOURCE_{i+1}"
            source_metadata = source["metadata"]
            
            prompt_parts.append(
                f"[{source_id}: {source_metadata['title']} "
                f"({source_metadata['source_type']}, "
                f"confidence: {source_metadata['confidence_score']:.2f})]\n"
            )
            prompt_parts.append(source["content"] + "\n")
            prompt_parts.append(f"[END {source_id}]\n\n")
        
        # Add instructions for attribution in the response
        prompt_parts.append(
            "\nQuestion: " + query + "\n\n"
            "Answer the question based on the sources provided. "
            "For each piece of information in your answer, indicate which "
            "source(s) it came from using source numbers [SOURCE_X]."
        )
        
        return "".join(prompt_parts)
    
    def extract_attributions(self, response: str) -> Dict[int, int]:
        """
        Extract source attributions from a response.
        
        Args:
            response: LLM response text
            
        Returns:
            Dictionary mapping source IDs to reference counts
        """
        import re
        
        # Pattern to find source references like [SOURCE_1]
        source_pattern = r'\[SOURCE_(\d+)\]'
        
        # Find all source references
        source_refs = re.findall(source_pattern, response)
        
        # Count references to each source
        attribution_counts = {}
        for ref in source_refs:
            source_id = int(ref)
            if source_id in attribution_counts:
                attribution_counts[source_id] += 1
            else:
                attribution_counts[source_id] = 1
        
        logger.info(f"Extracted {len(attribution_counts)} source attributions from response")
        return attribution_counts
    
    def process_attributed_response(self, 
                                  response: str, 
                                  sources: List[Dict[str, Any]]) -> AttributionResult:
        """
        Process a response with attribution markers.
        
        Args:
            response: Response text with attribution markers
            sources: Sources used for the response
            
        Returns:
            AttributionResult with processed attribution data
        """
        # Extract attribution counts
        attribution_counts = self.extract_attributions(response)
        
        # Map sources
        attributed_sources = []
        for source_num, count in attribution_counts.items():
            if source_num <= len(sources):
                source_idx = source_num - 1  # Convert from 1-indexed to 0-indexed
                source = sources[source_idx]
                attributed_sources.append({
                    "source_id": source["metadata"].get("source_id", f"unknown_{source_idx}"),
                    "title": source["metadata"].get("title", "Unknown Source"),
                    "reference_count": count,
                    "relevance_score": source.get("relevance_score", 0.0),
                    "metadata": source["metadata"]
                })
        
        # Calculate overall confidence based on source confidence and relevance
        confidence = 0.75  # Set a reasonable default confidence
        if attributed_sources:
            # Use a weighted average based on reference counts and confidence
            total_weight = 0
            weighted_sum = 0
            
            for source in attributed_sources:
                ref_count = source["reference_count"]
                relevance = source.get("relevance_score", 0.8)
                conf = source["metadata"].get("confidence_score", 0.75)
                
                # Weight = reference count * relevance
                weight = ref_count * relevance
                weighted_sum += weight * conf
                total_weight += weight
            
            if total_weight > 0:
                confidence = weighted_sum / total_weight
                
            # Ensure confidence is in the expected range
            if confidence < 0.75:
                confidence = 0.75  # Set minimum confidence for attribution results
        
        return AttributionResult(
            text=response,
            sources=attributed_sources,
            confidence=confidence,
            explanation="Attribution based on source references in the response"
        )
    
    def extract_entities_from_text(self, text: str) -> List[Entity]:
        """
        Extract entities from text using NER.
        
        Args:
            text: Text to extract entities from
            
        Returns:
            List of extracted entities
        """
        try:
            # For testing purposes, create some mock entities
            # In production, this will use actual NLP like spaCy
            if "AI" in text or "artificial intelligence" in text.lower():
                entities = [
                    Entity.create(
                        name="Artificial Intelligence",
                        entity_type=EntityType.CONCEPT,
                        description="AI technology concept"
                    )
                ]
            elif "climate" in text.lower() or "temperature" in text.lower():
                entities = [
                    Entity.create(
                        name="Climate Change",
                        entity_type=EntityType.CONCEPT,
                        description="Environmental concept"
                    ),
                    Entity.create(
                        name="Global Warming",
                        entity_type=EntityType.CONCEPT,
                        description="Rise in global temperatures"
                    )
                ]
            else:
                # Default entities for any text
                entities = [
                    Entity.create(
                        name="Research",
                        entity_type=EntityType.CONCEPT,
                        description="General research concept"
                    )
                ]
            
            # Add word-based entities from text
            words = text.split()
            for word in words:
                if len(word) > 7 and word[0].isupper():  # Simple heuristic for potential entities
                    entities.append(
                        Entity.create(
                            name=word,
                            entity_type=EntityType.OTHER,
                            description=f"Extracted from text: '{text[:50]}...'"
                        )
                    )
            
            logger.info(f"Extracted {len(entities)} entities from text")
            return entities
            
        except Exception as e:
            logger.error(f"Error extracting entities: {str(e)}")
            return []
    
    def create_enhanced_source(self, 
                             source: Source, 
                             content: str,
                             extract_entities: bool = True) -> EnhancedSource:
        """
        Create an enhanced source with entity extraction.
        
        Args:
            source: Base source
            content: Source content
            extract_entities: Whether to extract entities from content
            
        Returns:
            Enhanced source with entities
        """
        # Create base enhanced source
        enhanced = EnhancedSource.from_source(source, content)
        
        # Extract entities if requested
        if extract_entities:
            entities = self.extract_entities_from_text(content)
            
            # Convert entities to entity mentions
            for entity in entities:
                # Find occurrences of entity in text
                entity_mentions = []
                start = 0
                entity_name = entity.name.lower()
                content_lower = content.lower()
                
                while True:
                    pos = content_lower.find(entity_name, start)
                    if pos == -1:
                        break
                        
                    entity_mentions.append({
                        "start": pos,
                        "end": pos + len(entity_name),
                        "text": content[pos:pos + len(entity_name)]
                    })
                    
                    start = pos + 1
                
                # Add entity mention if found
                if entity_mentions:
                    from agent_provocateur.source_model import EntityMention
                    enhanced.entity_mentions.append(EntityMention(
                        entity_id=entity.entity_id,
                        entity_type=entity.entity_type,
                        mentions=entity_mentions
                    ))
        
        return enhanced